File: jEverest/src/ca/marc/everest/datatypes/ST.java
/* 
 * Copyright 2008-2011 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * User: Justin Fyfe
 * Date: 08-30-2011
 */
package ca.marc.everest.datatypes;

import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Locale;

import ca.marc.everest.annotations.*;
import ca.marc.everest.datatypes.generic.*;
import ca.marc.everest.datatypes.interfaces.IAny;
import ca.marc.everest.datatypes.interfaces.IPredicate;
import ca.marc.everest.interfaces.IResultDetail;
import ca.marc.everest.interfaces.ResultDetailType;
import ca.marc.everest.resultdetails.DatatypeValidationResultDetail;

/**
 * The character string data type is used to encapsulate text data that 
 * is primarily intended for machine use such as sorting, indexing and searching 
 */
@Structure(name = "ST", structureType = StructureType.DATATYPE)
public class ST extends PDV<String> {
	
	// backing field for language property
	private String m_language;
	// backing field for translations
	private SET<ST> m_translation;
	
	/**
	 * Creates a new empty instance of the ST class with the default language code
	 */
	public ST() { 
		super();
		this.m_language = Locale.getDefault().getLanguage() + "-" + Locale.getDefault().getCountry();
	}
	/**
	 * Creates a new instance of the ST class with the specified text in the 
	 * current locale's language code
	 */
	public ST(String data) {
		this();
		this.setValue(data);
	}
	/**
	 * Creates a new instance of the ST class with the specified string data 
	 * and language code
	 */
	public ST(String data, String language)
	{
		this(data);
		this.m_language = language;
	}
	
	/**
	 * Gets the current language of the character string data
	 */
	@Property(name = "language", propertyType = PropertyType.STRUCTURAL, conformance = ConformanceType.REQUIRED)
	public String getLanguage() { return this.m_language; }
	/**
	 * Sets the language of the character string data
	 */
	public void setLanguage(String value) { this.m_language = value; }
	/**
	 * Gets a set that contains translations of this string instance to other languages
	 */
	@Property(name = "translation", propertyType = PropertyType.NONSTRUCTURAL, conformance = ConformanceType.REQUIRED)
	public SET<ST> getTranslation() { return this.m_translation; }
	/**
	 * Sets a set that contains translations of this string instance to other languages.
	 */
	public void setTranslation(SET<ST> value) { this.m_translation = value; }
	/**
	 * Gets the length of this character string. 0 if the value has not been set
	 */
	public int length() { return this.getValue() == null ? 0 : this.getValue().length(); }
	/**
	 * Concatenate this string with another
	 */
	public ST concat(ST other)
	{
		ST retVal = new ST();
		if(other == null)
			return null;
		else if(other.isNull() || this.isNull())
			retVal.setNullFlavor(NullFlavor.NoInformation);
		else
			retVal.setValue(this.getValue() + other.getValue());
		return retVal;
	}
	/**
	 * Return a sub string of this character string instance
	 */
	public ST subString(INT start, INT end)
	{
		ST retVal = new ST();
		if(start == null || start.isNull())
			throw new IllegalArgumentException("start is null or null flavored");
		else if(end == null || end.isNull())
			throw new IllegalArgumentException("end is null or null flavored");
		else if(start.getValue() > this.length())
			throw new ArrayIndexOutOfBoundsException("start must be less than length of string");
		else if(end.getValue() > this.length())
			throw new ArrayIndexOutOfBoundsException("end must be less than length of the string");
		else if(start.compareTo(end) > 0)
			throw new IllegalArgumentException("start must be less than end");
		else if(this.getNullFlavor() != null)
			retVal.setNullFlavor(this.getNullFlavor());
		else
			retVal.setValue(this.getValue() == null ? null : this.getValue().substring(start.getValue(), end.getValue()));
		return retVal;
	}
	
	/**
	 * Convert this ST to an ED
	 * @return
	 * @throws UnsupportedEncodingException 
	 */
	public ED toED() throws UnsupportedEncodingException
	{
        // Copy
        ED retVal = new ED(this.getValue(), this.getLanguage());
        retVal.setControlActExt(this.getControlActExt());
        retVal.setControlActRoot(this.getControlActRoot());
        retVal.setFlavorId(this.getFlavorId());
        retVal.setNullFlavor(this.getNullFlavor());
        retVal.setTranslation(new SET<ED>(this.getTranslation()));
        retVal.setUpdateMode(this.getUpdateMode());
        retVal.setValidTimeHigh(this.getValidTimeHigh());
        retVal.setValidTimeLow(this.getValidTimeLow());
        retVal.setMediaType("text/plain");
        retVal.setRepresentation(EncapsulatedDataRepresentation.Text);
        return retVal;
	}
	
	/**
	 * Validates this instance of ST meets basic validation criteria
	 */
	@Override
	public boolean validate()
	{
		boolean isValid = ((this.getValue() != null) ^ (this.getNullFlavor() != null));
		if(this.getTranslation() != null)
			for(ST trans : this.getTranslation())
				isValid &= trans.getTranslation() == null;
		return isValid;
	}
	
	/**
	 * Validates this ST instance meets validation criteria for flavor
	 * NoTranslations
	 */
	@Flavor(name = "ST.NT")
	public boolean isValidNoTranslationsFlavor(ST st)
	{
		return st.getTranslation() == null || st.getTranslation().isEmpty();
	}
	
	/**
	 * Validates that this ST instance meets validation criteria for flavor
	 * Simple
	 */
	@Flavor(name = "ST.SIMPLE")
	public boolean isValidSimpleFlavor(ST st)
	{
		String defaultLanguage = Locale.getDefault().getLanguage() + "-" + Locale.getDefault().getCountry();
		// Language is usually set to the default locale
        if (st.getLanguage().equals(defaultLanguage))
            st.setLanguage(null);
        return st.getLanguage() == null;
	}
	/**
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = super.hashCode();
		result = prime * result
				+ ((m_language == null) ? 0 : m_language.hashCode());
		result = prime * result
				+ ((m_translation == null) ? 0 : m_translation.hashCode());
		return result;
	}
	/**
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (!super.equals(obj))
			return false;
		if (getClass() != obj.getClass())
			return false;
		ST other = (ST) obj;
		if (m_language == null) {
			if (other.m_language != null)
				return false;
		} else if (!m_language.equals(other.m_language))
			return false;
		if (m_translation == null) {
			if (other.m_translation != null)
				return false;
		} else if (!m_translation.equals(other.m_translation))
			return false;
		return true;
	}
	/**
	 * @see ca.marc.everest.datatypes.generic.PDV#validateEx()
	 */
	@Override
	public Collection<IResultDetail> validateEx() {
		 List<IResultDetail> retVal = (ArrayList<IResultDetail>)super.validateEx();

		 // A predicate which can detect if a ST has a null translation
		 IPredicate<ST> nullTranslationPredicate = new IPredicate<ST>()
         {
			 public boolean match(ST o)
			 {
				 return o.getTranslation() != null;
			 }
		 };
		 
         if(this.isNull() && this.getValue() != null)
             retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "ST", EverestValidationMessages.MSG_NULLFLAVOR_WITH_VALUE, null));
         else if(this.getValue() == null && !this.isNull())
             retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "ST", EverestValidationMessages.MSG_NULLFLAVOR_MISSING, null));
         if (this.getTranslation() != null && this.getTranslation().findAll(nullTranslationPredicate).size() == this.getTranslation().size())
             retVal.add(new DatatypeValidationResultDetail(ResultDetailType.ERROR, "ST", String.format(EverestValidationMessages.MSG_PROPERTY_NOT_PERMITTED, "Translation", "Translation"), null));
         return retVal;
	}
	/**
	 * @see ca.marc.everest.datatypes.ANY#semanticEquals(ca.marc.everest.datatypes.interfaces.IAny)
	 */
	@Override
	public BL semanticEquals(IAny other) {
		
		BL retVal = new BL(false);
		if (other == null)
            return null;
        else if (this.isNull() && other.isNull())
            retVal.setNullFlavor(this.getNullFlavor().getCode().getCommonParent(other.getNullFlavor().getCode()));
        else if (this.isNull() ^ other.isNull())
            retVal.setNullFlavor(NullFlavor.NotApplicable);
        else
        {
        	if(other instanceof ST)
        	{
        		ST otherSt = (ST)other;
        		retVal.setValue(otherSt.getValue() != null ? otherSt.getValue().equals(this.getValue()) : this.getValue() == null);
        	}
        	else if(other instanceof ED)
        	{
        		ED otherEd = (ED)other; // ST may also be equal to ED when value is text/plain
        		if (otherEd.getMediaType() == "text/plain")
        			retVal.setValue(otherEd.getData() != null ? otherEd.getData().equals(this.getValue().getBytes()) : this.getValue() == null);
        		else
        			retVal.setValue(false);
        	}
        }
        return retVal;
	}
	
	
}

File: MARC.Everest/DataTypes/ST.cs
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.

 * 
 * User: Justin Fyfe
 * Date: 01-09-2009
 */
using System;
using System.Collections.Generic;
using System.Text;
using MARC.Everest.Attributes;
using System.Globalization;
using System.Xml.Serialization;
using MARC.Everest.DataTypes.Interfaces;
using MARC.Everest.Connectors;

namespace MARC.Everest.DataTypes
{
    /// <summary>
    /// The character string datatype stands for text data primarily intended for machine processing
    /// </summary>
    /// <example>Using the <see cref="T:MARC.Everest.DataTypes.ST"/> class
    /// <code lang="cs" title="ST datatype">
    /// <![CDATA[
    /// // String Hello with current language code 
    /// ST test = new ST();
    /// test.Value = "Hello!";
    /// // String hello! with current language code
    /// test = "Hello!";
    /// // String hello! in a different language
    /// test = new ST("Bonjour!", "fr-ca");
    /// ]]>
    /// </code>
    /// </example>
    [Structure(Name = "ST", StructureType = StructureAttribute.StructureAttributeType.DataType)]
    [XmlType("ST", Namespace = "urn:hl7-org:v3")]
#if !WINDOWS_PHONE
    [Serializable]
#endif
    public class ST : PDV<String>, IEquatable<ST>
    {

        /// <summary>
        /// Create a new instance of the <see cref="T:MARC.Everest.DataTypes.ST"/> type
        /// </summary>
        public ST() : base() { }
        /// <summary>
        /// Create a new instance of the <see cref="T:MARC.Everest.DataTypes.ST"/> type using <paramref name="data"/> as the initial contents
        /// </summary>
        /// <param name="data">The initial data</param>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2214:DoNotCallOverridableMethodsInConstructors")]
        public ST(string data) : this() { this.Value = data; }
        /// <summary>
        /// Create a new instance of the <see cref="T:MARC.Everest.DataTypes.ST"/> type using <paramref name="data"/> as the initial contents with 
        /// <paramref name="language"/>
        /// </summary>
        /// <param name="data">The intial data</param>
        /// <param name="language">The language the data is represented in</param>
        public ST(string data, string language) : this(data) { this.Language = language; }

        /// <summary>
        /// The human language of the content
        /// </summary>
        [Property(Name = "language", PropertyType = PropertyAttribute.AttributeAttributeType.Structural, Conformance = PropertyAttribute.AttributeConformanceType.Optional)]
        public string Language { get; set; }

        /// <summary>
        /// Alternative renditions of the same content translated into a different language
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly"), Property(Name = "translation", PropertyType = PropertyAttribute.AttributeAttributeType.NonStructural, Conformance = PropertyAttribute.AttributeConformanceType.Optional)]
        public SET<ST> Translation { get; set; }

        #region Operations

        /// <summary>
        /// Gets the length of the string
        /// </summary>
        public int Length { get { return this.Value != null ? this.Value.Length : 0; } }

        /// <summary>
        /// Concatenate this string value with <paramref name="other"/>
        /// </summary>
        public ST Concat(ST other)
        {
            if (other == null)
                return null;
            else if (other.IsNull && this.IsNull)
                return new ST() { NullFlavor = this.NullFlavor.GetCommonParent(other.NullFlavor) };
            else if (other.IsNull)
                return new ST() { NullFlavor = MARC.Everest.DataTypes.NullFlavor.NoInformation };
            else
                return new ST(this.Value + other.Value);
        }

        /// <summary>
        /// Performs a substring on the <see cref="T:MARC.Everest.DataTypes.ST"/> as per HL7v3 datatypes rules
        /// </summary>
        /// <exception cref="T:ArgumentNullException">When either <paramref name="start"/> or <paramref name="end"/> are null or nullFlavored</exception>
        /// <exception cref="T:ArgumentOutOfRangeException">When <paramref name="start"/> is greater than <paramref name="end"/> or either <paramref name="end"/> or <paramref name="start"/> are
        /// beyond the length of the string</exception>
        public ST Substring(INT start, INT end)
        {
            if (start == null || start.IsNull)
                throw new ArgumentNullException("start", "Argument must not be null and must not contain a null flavor");
            else if (end == null || end.IsNull)
                throw new ArgumentNullException("end", "Argument must not be null and must not contain a null flavor");
            else if (start > this.Length)
                throw new ArgumentOutOfRangeException("start", "Value must be less than the size of the string");
            else if (end > this.Length)
                throw new ArgumentOutOfRangeException("end", "Value must be less than the size of the string");
            else if (start > end)
                throw new ArgumentOutOfRangeException("start", "Value must be less than end");
            else if(this.NullFlavor != null)
                return new ST() { NullFlavor = this.NullFlavor };
            return new ST(this.Value == null ? "" : this.Value.Substring((int)start, (int)(end - start)));
        }

        #endregion

        #region Operators

        /// <summary>
        /// Concatenation operator +
        /// </summary>
        public static ST operator +(ST a, ST b)
        {
            if (a != null)
                return a.Concat(b);
            else
                throw new ArgumentNullException("Cannot concatenate null string value");
        }

        /// <summary>
        /// Converts a <see cref="T:ST"/> to a <see cref="T:System.String"/> in a safe way
        /// </summary>
        /// <param name="o">The <see cref="T:MARC.Everest.DataTypes.ST"/> to convert</param>
        /// <returns>The converted string</returns>
        /// <remarks>The implicit cast operator will convert only if <see cref="T:MARC.Everest.DataTypes.ST"/> is not null. This differs from the explicit cast where 
        /// no exception is thrown</remarks>
        /// <exception cref="T:System.InvalidCastException">When the <see cref="T:MARC.Everest.DataTypes.ST"/> instance has a nullflavor</exception>
        public static implicit operator String(ST o)
        {
            if (o == null || o.IsNull)
                return null;
            else
                return o.Value;
        }
        
        /// <summary>
        /// Casts a <see cref="T:MARC.Everest.DataTypes.Ed"/> into an <see cref="T:ST"/>
        /// </summary>
        /// <param name="o">The object to cast</param>
        /// <returns>The cast <see cref="T:MARC.Everest.DataTypes.ST"/></returns>
        public static implicit operator ST(ED o)
        {
            if (o == null) return null;

#if WINDOWS_PHONE
            string value = o.Representation == MARC.Everest.DataTypes.Interfaces.EncapsulatedDataRepresentation.B64 ?
                o.Base64Data : o.Representation == MARC.Everest.DataTypes.Interfaces.EncapsulatedDataRepresentation.XML ?
                o.Value : o.Value;
#else
            string value = o.Representation == MARC.Everest.DataTypes.Interfaces.EncapsulatedDataRepresentation.B64 ?
                o.Base64Data : o.Representation == MARC.Everest.DataTypes.Interfaces.EncapsulatedDataRepresentation.XML ? 
                o.XmlData.OuterXml : o.Value;
#endif
            return new ST(value)
            {
                ControlActExt = o.ControlActExt,
                ControlActRoot = o.ControlActRoot,
                Flavor = o.Flavor,
                Language = o.Language,
                NullFlavor = o.NullFlavor,
                UpdateMode = o.UpdateMode,
                ValidTimeHigh = o.ValidTimeHigh,
                ValidTimeLow = o.ValidTimeLow
            };
                
        }

        /// <summary>
        /// Convert a ST into an ED
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "o")]
        internal ED ToEDUtil(ST o)
        {
            if (o == null)
                return null;

            // Copy
            ED retVal = new ED(o.Value, o.Language);
            retVal.ControlActExt = o.ControlActExt;
            retVal.ControlActRoot = o.ControlActRoot;
            retVal.Flavor = o.Flavor;
            retVal.NullFlavor = o.NullFlavor;
            retVal.Translation = new SET<ED>(o.Translation);
            retVal.UpdateMode = o.UpdateMode;
            retVal.ValidTimeHigh = o.ValidTimeHigh;
            retVal.ValidTimeLow = o.ValidTimeLow;
            retVal.MediaType = "text/plain";
            retVal.Representation = EncapsulatedDataRepresentation.TXT;
            return retVal;
        }

        /// <summary>
        /// Converts a <see cref="String"/> to a <see cref="ST"/>
        /// </summary>
        /// <param name="o">String to convert</param>
        /// <returns>Converted <see cref="T:MARC.Everest.DataTypes.ST"/></returns>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2225:OperatorOverloadsHaveNamedAlternates"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "o")]
        public static implicit operator ST(String o)
        {
            if (o == null)
                return null;

            ST retVal = new ST();
            retVal.Language = CultureInfo.CurrentCulture.Name; 
            retVal.Value = o;
            return retVal;
        }

        /// <summary>
        /// Converts a <see cref="T:MARC.Everest.DataTypes.ST"/> to a <see cref="T:MARC.Everest.DataTypes.REAL"/>
        /// </summary>
        /// <param name="o">ST to convert</param>
        /// <returns>Converted REAL</returns>
        /// <remarks>The resulting REAL will have a nullflavor of NoInformation if the cast was not possible. 
        /// <para>The result will also have a precision set based on the location of the decimal point in the string representation</para></remarks>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2225:OperatorOverloadsHaveNamedAlternates"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "o")]
        public static explicit operator REAL(ST o)
        {
            if (o == null)
                return null;

            Double d = 0;
            if (!Double.TryParse(o.Value, out d))
                return new REAL() { NullFlavor = DataTypes.NullFlavor.NoInformation };

            int precision = 0;
            if (o.Value.Contains("."))
                precision = o.Value.Length - o.Value.IndexOf(".") - 1;

            return new REAL(d)
            {
                Precision = precision
            };
        }

        /// <summary>
        /// Converts a <see cref="T:MARC.Everest.DataTypes.ST"/> to a <see cref="T:MARC.Everest.DataTypes.INT"/>
        /// </summary>
        /// <param name="o">ST to convert</param>
        /// <returns>Converted INT</returns>
        /// <remarks>The resulting INT will have a nullflavor of NoInformation if the cast was not possible. </remarks>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA2225:OperatorOverloadsHaveNamedAlternates"), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly", MessageId = "o")]
        public static explicit operator INT(ST o)
        {
            if (o == null)
                return null;

            int d = 0;
            if (!Int32.TryParse(o.Value, out d))
                return new INT() { NullFlavor = DataTypes.NullFlavor.NoInformation };

            return new INT(d);
        }

        #endregion


        #region IEquatable<ST> Members

        /// <summary>
        /// Determines if this ST equals another ST
        /// </summary>
        public bool Equals(ST other)
        {
            bool result = false;
            if (other != null)
                result = base.Equals((PDV<string>)other) &&
                    other.Language == this.Language &&
                    (other.Translation != null ? other.Translation.Equals(this.Translation) : this.Translation == null);
            return result;
        }

        /// <summary>
        /// Override of base equals
        /// </summary>
        public override bool Equals(object obj)
        {
            if (obj is ST || obj is string)
                return Equals(obj as ST);
            return base.Equals(obj);
        }

        #endregion

        /// <summary>
        /// Validate the <see cref="T:MARC.Everest.DataTypes.ST"/> instance
        /// </summary>
        /// <remarks>
        /// An <see cref="T:MARC.Everest.DataTypes.ST"/> is considered valid if:
        /// <list type="bullet">
        ///     <item><see cref="P:Value"/>Value is populated XOR <see cref="P:NullFlavor"/> is populated, and</item>
        ///     <item>No value in the <see cref="P:Translation"/> contains a translation</item>
        /// </list>
        /// </remarks>
        public override bool Validate()
        {
            return ((this.Value != null) ^ (this.NullFlavor != null)) &&
                (this.Translation != null && this.Translation.FindAll(o => o.Translation == null).Count == this.Translation.Count ||
                this.Translation == null);
        }

        /// <summary>
        /// Extended validation function that returns detected issues
        /// </summary>
        public override IEnumerable<Connectors.IResultDetail> ValidateEx()
        {
            var retVal = base.ValidateEx() as List<IResultDetail>;

            if(this.NullFlavor != null && this.Value != null)
                retVal.Add(new DatatypeValidationResultDetail(ResultDetailType.Error, "ST", ValidationMessages.MSG_NULLFLAVOR_WITH_VALUE, null));
            else if(this.Value == null && this.NullFlavor == null)
                retVal.Add(new DatatypeValidationResultDetail(ResultDetailType.Error, "ST", ValidationMessages.MSG_NULLFLAVOR_MISSING, null));
            if (this.Translation != null && this.Translation.FindAll(o => o.Translation != null).Count == this.Translation.Count)
                retVal.Add(new DatatypeValidationResultDetail(ResultDetailType.Error, "ST", String.Format(ValidationMessages.MSG_PROPERTY_NOT_PERMITTED_ON_PROPERTY, "Translation", "Translation"), null));
            return retVal;
        }

        /// <summary>
        /// ST.NT (no translations) flavor
        /// </summary>
        [Flavor("NT")]
        [Flavor("ST.NT")]
        public static bool IsValidNtFlavor(ST st)
        {
            return st.Translation == null || st.Translation.IsEmpty;
        }

        /// <summary>
        /// Creates an instance of the <see cref="T:MARC.Everest.DataTypes.ST"/> type such that it is conformant to the NoTranslations 
        /// flavor
        /// </summary>
        /// <param name="value">The value of the string</param>
        /// <returns>The constructed <see cref="T:MARC.Everest.DataTypes.ST"/> conformant to the NT flavor</returns>
        public static ST CreateNt(string value)
        {
            return new ST(value) { Translation = null };
        }

        /// <summary>
        /// ST.SIMPLE contains no language
        /// </summary>
        [Flavor("SIMPLE")]
        [Flavor("ST.SIMPLE")]
        public static bool IsValidSimpleFlavor(ST st)
        {
            // Language is usually set to the default locale
            if (st.Language == CultureInfo.CurrentCulture.Name)
                st.Language = null;
            return st.Language == null;
        }

        /// <summary>
        /// Creates an instance of <see cref="T:MARC.Everest.DataTypes.ST"/> type that is conformant to the SIMPLE flavor
        /// </summary>
        /// <param name="value">The value of the <see cref="T:MARC.Everest.DataTypes.ST"/></param>
        /// <returns>An instance of <see cref="T:MARC.Everest.DataTypes.ST"/> conformant to the SIMPLE flavor</returns>
        public static ST CreateSimple(string value)
        {
            return new ST(value) { Language = null };
        }

        /// <summary>
        /// Semantic equals
        /// </summary>
        public override BL SemanticEquals(IAny other)
        {
            if (other == null)
                return null;
            else if (this.IsNull && other.IsNull)
                return new BL() { NullFlavor = NullFlavorUtil.GetCommonParent(this.NullFlavor, other.NullFlavor) };
            else if (this.IsNull ^ other.IsNull)
                return new BL() { NullFlavor = DataTypes.NullFlavor.NotApplicable };

            ST otherSt = other as ST;
            if(otherSt != null)
                return otherSt.Value != null ? otherSt.Value.Equals(this.Value) : this.Value == null;
            ED otherEd = other as ED; // ST may also be equal to ED when value is text/plain
            if (otherEd.MediaType == "text/plain")
                return otherEd.Value != null ? otherEd.Value.Equals(this.Value) : this.Value == null;
            return false;
        }
    }
}
